### DISCUSSION TITLE
About Extending WebUI Pages in the Pyri Framework

### DISCUSSION THREAD

**Discussion by Lafatalite** [pyri-project/pyri-core#12] [Q&A]
I want to add a new page to the webui. I have already read the documentation but didn't understand it. I tried to emulate the robotics_browser project and packaged it into a wheel file, then copied it to ./local/share/pyri-webui-server, but it didn't display successfully. I'm wondering if there is a simpler example, like a page that displays "Hello World", to understand its working principle. The robotics_browser seems very complex, and I also want to know the difference between robotics_browser and robotics. From my understanding, projects with the "browser" suffix are webui projects that include a frontend page and have some backend functionalities. Are they interconnected, or do they need to be independent?

**Comment by Lafatalite** [comment]
I just tried deploying the pyri-example-plugin. I packaged the project into a wheel and then copied it to ./local/share/pyri-webui-server and then restarted the entire pyri, but I still can't see the new content. Is there something wrong with the deployment steps?

**Comment by Lafatalite** [comment]
I just tested it again and found that both blockly and sandbox in the example-plugin have been successfully loaded and run, but there isn't a webui_server page. I see there is a hello.html in the directory but it doesn't appear. Is there something wrong?

**Comment by johnwason** [comment]
Pyri has two major execution environments: the "runtime" side, which contains the nodes that run as normal Python programs, and the "webui" which runs inside the web browser using Pyodide. The `pyri-robotics` package is on the "runtime" side, and contains plugins and nodes that add the functionality for robotics. This functionality includes adding sandbox functions/blockly blocks, and the jog/motion nodes (running in their own programs) that implement some of the functionality.

The [Pyodide](https://github.com/pyodide/pyodide) webui side is very different. This uses WebAssembly to run Python inside the web browser itself. The user interface uses a combination of Pyodide, Vue, Bootstrap, and other JavaScript components to implement the user interface. The Python code and resources are stored in Python wheels, and the wheels are placed in a special directory `$HOME/.local/pyri-webui-server/wheels`. These wheels are loaded into Pyodide at start, and the plugin factories are used to initialize the panels that are displayed in the webui. You can use [pyri-cli](https://github.com/pyri-project/pyri-cli) "dev" commands to automatically build your wheel and place it in the correct directory to be loaded by Pyodide.

Let's look at the some important points for [pyri-robotics-browser](https://github.com/pyri-project/pyri-robotics-browser), which is loaded into the webui and provides the "Jog" panel.

- The [pyproject.toml](https://github.com/pyri-project/pyri-robotics-browser/blob/master/pyproject.toml) must contain several important entries:
  - The [keywords](https://github.com/pyri-project/pyri-robotics-browser/blob/549fecff9b8b291be2a2f499a9f22eecf5fdc069/pyproject.toml#L9) entry must contain `pyri-webui-browser-package`, `pyodide`, and `pyri-webui`. These keywords are used by `pyri-cli` and the rest of the system to detect the runtime/webui packages
  - The package generally will [depend](https://github.com/pyri-project/pyri-robotics-browser/blob/549fecff9b8b291be2a2f499a9f22eecf5fdc069/pyproject.toml#L19) on `pyri-common` and `improtlib-resources`.
  - The [build-system](https://github.com/pyri-project/pyri-robotics-browser/blob/549fecff9b8b291be2a2f499a9f22eecf5fdc069/pyproject.toml#L34) and [tool.setuptools.package-data](https://github.com/pyri-project/pyri-robotics-browser/blob/549fecff9b8b291be2a2f499a9f22eecf5fdc069/pyproject.toml#L41C2-L41C30) must be configured to include any non-python files you need, such as html, js, svg, jpg, etc files you may need to include in the wheel.
  - The entry-points for [panels](https://github.com/pyri-project/pyri-robotics-browser/blob/549fecff9b8b291be2a2f499a9f22eecf5fdc069/pyproject.toml#L28) and [vue components](https://github.com/pyri-project/pyri-robotics-browser/blob/549fecff9b8b291be2a2f499a9f22eecf5fdc069/pyproject.toml#L31) must be correctly configured. These entry-points are used to direct the webui what code to load from the wheel.
- The factories must be implemented:
  - [Vue Components](https://github.com/pyri-project/pyri-robotics-browser/blob/master/src/pyri/robotics_browser/components/robotics_components.py)
  - [Panels](https://github.com/pyri-project/pyri-robotics-browser/blob/master/src/pyri/robotics_browser/panels/robotics_panels.py)
- Define components. The PyRI WebUI uses Vue, and a compatibility layer was developed to allow for easy use of Vue with Python. If you look at the [jog_panel.py](https://github.com/pyri-project/pyri-robotics-browser/blob/master/src/pyri/robotics_browser/components/jog_component.py) component, the `@vue_method` and other `@vue_*` decorators are used to expose the Python to Vue. The adapter is defined in [pyri-webui-browser vue.py](https://github.com/pyri-project/pyri-robotics-browser/blob/master/src/pyri/robotics_browser/components/jog_component.py). Looking at [jog_panel.html](https://github.com/pyri-project/pyri-robotics-browser/blob/master/src/pyri/robotics_browser/components/jog_component.html), we can see the visual design of the Vue component, which calls these methods. The tags starting with "b-" are [bootstrap-vue](https://bootstrap-vue.org/) components.
- Add a panel. In this case, the Jog panel will just initialize the jog component, in this case the `component_type="pyri-robotics-jog"` is how the panel specifies the component. See [Here](https://github.com/pyri-project/pyri-robotics-browser/blob/549fecff9b8b291be2a2f499a9f22eecf5fdc069/src/pyri/robotics_browser/panels/robotics_panels.py#L6)

Hope this helps. I think I will need to create more detailed tutorials in the future.



Note that the webui-routes are different, and are for adding HTTP routes to the webui server. This is independent of the Pyodide webui wheels, and is intended for adding resources and additional HTTP features. For your use case you most likely won't need this, unless you want to add JavaScript packages.

**Reply by Lafatalite** [reply]
I tried to emulate the Robotic Browser and wrote a "Hello World" panels, and also composed other required files according to your instructions. However, after packaging and placing into the default directory, I wasn't able to access the website which was stuck loading on the logo. When I remove the packaged wheel, the page loads. Did I make a mistake somewhere? Can Pyri enable logging? I have not found this option. These are the files I wrote.
[helloworld.tar.gz](https://github.com/pyri-project/pyri-core/files/15036223/helloworld.tar.gz)

**Reply by johnwason** [reply]
Look at the Javascript developer console to see the error message.

**Reply by Lafatalite** [reply]
![image](https://github.com/pyri-project/pyri-core/assets/35282465/e5aa6d73-8ad2-4371-8699-cae2e3d7799a)

**Reply by johnwason** [reply]
I don't see any directories in your tar file; you will need the src/pyri_example_plugin directory, and an __init__.py file in the directory along with your py/html files. You also don't need the sandbox and blockly entry points since they are not used in the webui.

**Reply by Lafatalite** [reply]
[pyri-example-plugin.tar.gz](https://github.com/pyri-project/pyri-core/files/15036929/pyri-example-plugin.tar.gz)

**Reply by Lafatalite** [reply]
I'm sorry to disturb you, but I am still attempting to load the web UI. Currently, I am observing a JavaScript error in the browser console, which indicates an unknown element 'pyri-example-hel'. This tag is custom-defined by me and is written following the architecture used in the robotic-browser and vision-browser projects. The error suggests that the element is not registered.
I understand that the first parameter of the vue_register_component function should correspond to the 'component_type' value found in the _panel_infos dictionary within the appropriate xxxxx_panels.py file, and I have already passed the identical value. However, despite this, it is still generating an error, and I am unable to access the page.
<img width="557" alt="Snipaste_2024-04-23_16-55-16" src="https://github.com/pyri-project/pyri-core/assets/35282465/10e8aa33-1cda-4052-b047-a2562a9b3941">
<img width="634" alt="Snipaste_2024-04-23_16-55-38" src="https://github.com/pyri-project/pyri-core/assets/35282465/7a1fbf70-d792-4fcd-b7ca-167cce7a4070">
[pyri-example-plugin.tar.gz](https://github.com/pyri-project/pyri-core/files/15074112/pyri-example-plugin.tar.gz)

**Reply by Lafatalite** [reply]
<img width="911" alt="Snipaste_2024-04-23_16-56-01" src="https://github.com/pyri-project/pyri-core/assets/35282465/5be56f16-03e6-4263-9b90-4d36769cf151">

**Comment by johnwason** [comment]
Hi @Lafatalite I am going to put together an example repository, stand by...

**Comment by johnwason** [comment]
See https://github.com/pyri-project/pyri-example-browser
